shared/src/ledger/governance/mod.rs

```diff
@@ -0,0 +1,493 @@
+                    let min_funds_parameter_key =
+                        gov_storage::get_min_proposal_fund_key();
+                    let min_funds_parameter: Option<Amount> = read(
+                        &self.ctx,
+                        &min_funds_parameter_key,
+                        ReadType::PRE,
+                    )
+                    .ok();
+                    let pre_balance: Option<Amount> =
+                        read(&self.ctx, &balance_key, ReadType::PRE).ok();
+                    let post_balance: Option<Amount> =
+                        read(&self.ctx, &balance_key, ReadType::POST).ok();
+                    match (min_funds_parameter, pre_balance, post_balance) {
+                        (
+                            Some(min_funds_parameter),
+                            Some(pre_balance),
+                            Some(post_balance),
+                        ) => {
+                            post_balance > pre_balance
+                                && post_balance - pre_balance
+                                    >= min_funds_parameter
+                        }
+                        (
+                            Some(min_funds_parameter),
+                            None,
+                            Some(post_balance),
+                        ) => post_balance >= min_funds_parameter,
+                        _ => false,
+                    }
+                }
+                (KeyType::PARAMETER, _) => false,
+                (KeyType::UNKNOWN, _) => false,
+                _ => false,
+            }
+        });
+        Ok(result)
+    }
+}
+
+fn is_valid_key_set<DB, H, CA>(
+    context: &Ctx<DB, H, CA>,
+    keys: &BTreeSet<Key>,
+) -> bool
+where
+    DB: 'static + ledger_storage::DB + for<'iter> ledger_storage::DBIter<'iter>,
+    H: 'static + StorageHasher,
+    CA: 'static + WasmCacheAccess,
+{
+    if is_valid_proposal_init_key_set(context, keys) {
+        return true;
+    };
+    false
+}
+
+fn is_valid_proposal_init_key_set<DB, H, CA>(
+    context: &Ctx<DB, H, CA>,
+    keys: &BTreeSet<Key>,
+) -> bool
+where
+    DB: 'static + ledger_storage::DB + for<'iter> ledger_storage::DBIter<'iter>,
+    H: 'static + StorageHasher,
+    CA: 'static + WasmCacheAccess,
+{
+    let counter_key = gov_storage::get_counter_key();
+    let pre_counter: Option<u64> =
+        read(context, &counter_key, ReadType::PRE).ok();
+    if pre_counter.is_none() {
+        return false;
+    }
+    let counter = pre_counter.unwrap();
+
+    // Construct the set of expected keys
+    let mandatory_keys = BTreeSet::from([
+        counter_key,
+        gov_storage::get_content_key(counter),
+        gov_storage::get_author_key(counter),
+        gov_storage::get_funds_key(counter),
+        gov_storage::get_voting_start_epoch_key(counter),
+        gov_storage::get_voting_end_epoch_key(counter),
+    ]);
+
+    // Check that expected set is a subset the actual one
+    keys.is_superset(&mandatory_keys)
```

Comment:
one thing I'm not sure if this VP would handle correctly is submitting a transaction that creates multiple proposals, because only the first one (with the current counter value) would be validated by this function (with the whitelist trusted txs it's not such an issue, but with custom txs it could be)